# CHAPTER 2 : CRIME DISCOVERY

## Introduction
This chapter identifies the presence and structure of fraud within the transaction data. It begins by assessing the overall scale and distribution of fraudulent activity, then examines which transaction types are most associated with fraud. Finally, it analyzes temporal patterns to understand when fraud is most likely to occur. The goal is to establish clear evidence of how fraud manifests across volume, behavior, and time.

### Contents
- Q1: Scale and Distribution of Fraud
- Q2: Transaction Types Associated with Fraud
- Q3: Temporal Patterns of Fraudulent Activity

## Database Connection

In [3]:
# Establish connection to the MySQL database using the custom utility module
# Returns:
# - engine: used for running SQL queries
# - conn: active database connection
from db_connect import connect_to_db
engine, conn = connect_to_db()

Please enter database name:  fraud_analysis


   Using default: fraud_analysis


 Please enter the password for root@127.0.0.1:  ········



 The notebook is successfully connected to MYSQL!
  Server: 127.0.0.1:3306
  Database in use: fraud_analysis
  User: root
SQL magic configured - use %%sql in any cell!


## Q1 Scale and Distribution of Fraud

In [20]:
%%sql
SELECT 
    FORMAT(COUNT(*), 0) AS total_transactions,
    FORMAT(SUM(CASE WHEN isFraud = 1 THEN 1 ELSE 0 END), 0) AS total_fraud_transactions,
    CONCAT(ROUND(SUM(CASE WHEN isFraud = 1 THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2), '%') AS fraud_transaction_percentage,
    CONCAT('KSH ', FORMAT(SUM(CASE WHEN isFraud = 1 THEN amount ELSE 0 END), 2))  AS total_fraud_value,
    CONCAT(ROUND(SUM(CASE WHEN isFraud = 1 THEN amount ELSE 0 END) * 100.0 / SUM(amount), 2), '%') AS fraud_value_percentage
FROM paysim;

 * mysql+pymysql://root:***@127.0.0.1:3306/fraud_analysis
1 rows affected.


total_transactions,total_fraud_transactions,fraud_transaction_percentage,total_fraud_value,fraud_value_percentage
"6,362,620","8,213",0.13%,"KSH 12,056,415,427.84",1.05%


### Insight

The scale of fraud in the dataset is so minimal with low frequency but with the far greatest impact. With only 0.13% of total transactions, it accumulates to more than KSH 12 billion of financial losses. The fraud is majorly of low frequency but with strategic target of high value extraction. This is not random but organised exploitation of specific vulnerabilities in the system. Furthermore, The cost of false negative is so immense and catastrophic to the financial ecosystem compared to that of false positive where it majorly cause customer friction.

## Q2 Transaction types associated with fraud
Which transaction types are most associated with fraudulent activity, and what does this reveal about how fraudsters exploit specific transaction mechanisms within the platform?

In [28]:
%%sql
SELECT
    type AS transaction_type,
    FORMAT(COUNT(*), 0) AS total_transactions,
    SUM(CASE WHEN isFraud = 1 THEN 1 ELSE 0 END) AS total_fraud_transactions,
    CONCAT(ROUND(SUM(CASE WHEN isFraud = 1 THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2), '%') AS fraud_rate,
    CONCAT('KSH ', FORMAT(SUM(CASE WHEN isFraud = 1 THEN amount ELSE 0 END), 2)) AS total_fraud_amount,
    CONCAT(ROUND(SUM(CASE WHEN isFraud = 1 THEN amount ELSE 0 END) * 100.0 / SUM(amount), 2), '%') AS fraud_value_rate
FROM paysim
GROUP BY type
ORDER BY fraud_rate DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/fraud_analysis
5 rows affected.


transaction_type,total_transactions,total_fraud_transactions,fraud_rate,total_fraud_amount,fraud_value_rate
TRANSFER,"532,909",4097,0.77%,"KSH 6,067,213,184.01",1.25%
CASH_OUT,"2,237,500",4116,0.18%,"KSH 5,989,202,243.83",1.52%
PAYMENT,"2,151,495",0,0.00%,KSH 0.00,0.00%
DEBIT,"41,432",0,0.00%,KSH 0.00,0.00%
CASH_IN,"1,399,284",0,0.00%,KSH 0.00,0.00%


### Insight
Fraud in this system is not random it is tightly focused on just two transaction types that allow money to be moved and then withdrawn. PAYMENT, DEBIT, and CASH_IN show zero fraud because they don’t help a fraudster achieve their goal; they either bring money into the system or send it to traceable destinations. In contrast, TRANSFER is used to move and reposition funds, which is why it has the highest fraud rate, while CASH_OUT is where the fraud is actually completed, reflected in its slightly higher number of fraud cases and stronger value impact. This confirms that fraud follows a clear sequence: first move the money, then extract it.

## Q3 : Fraud occurence in daily, weekly and monthly
When does fraudulent activity most commonly occur across daily, weekly, and monthly time patterns, and what does this reveal about temporal            vulnerabilities in the platform's transaction system?

In [42]:
%%sql
SELECT
    'day' AS time_level,
    FLOOR(step/24) + 1 AS time_bucket,
    FORMAT(COUNT(*), 0) AS fraud_count,
    CONCAT('KSH ', FORMAT(SUM(amount), 2)) AS amount_lost
FROM paysim
WHERE
    isFraud = 1
GROUP BY time_level, time_bucket

UNION ALL

SELECT
    'week' AS time_level,
    FLOOR(step/168) + 1 AS time_bucket,
    FORMAT(COUNT(*), 0) AS fraud_count,
    CONCAT('KSH ', FORMAT(SUM(amount), 2)) AS amount_lost
FROM paysim
WHERE
    isFraud = 1
GROUP BY time_level, time_bucket

UNION ALL

SELECT
    'month' AS time_level,
    FLOOR(step/720) + 1 AS time_bucket,
    FORMAT(COUNT(*), 0) AS fraud_count,
    CONCAT('KSH ', FORMAT(SUM(amount), 2)) AS amount_lost
FROM paysim
WHERE
    isFraud = 1
GROUP BY time_level, time_bucket
ORDER BY time_level, fraud_count DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/fraud_analysis
38 rows affected.


time_level,time_bucket,fraud_count,amount_lost
day,17,316,"KSH 633,637,152.03"
day,3,306,"KSH 350,162,911.48"
day,2,305,"KSH 366,307,308.93"
day,12,296,"KSH 429,005,928.83"
day,31,282,"KSH 449,703,481.30"
day,26,280,"KSH 404,444,305.27"
day,7,280,"KSH 393,773,952.78"
day,27,276,"KSH 457,345,981.94"
day,24,276,"KSH 329,355,065.04"
day,10,274,"KSH 364,992,686.55"


### Insight 
Fraud in this system is not evenly distributed over time and shows clear peaks as well as variation in financial impact. The highest activity occurs on **Day 17**, with **316 fraud cases and about KSH 633 million lost**, making it both the peak in volume and value. However, fraud count does not consistently match money lost. For instance, **Day 29 (262 cases) results in KSH 515 million**, while **Day 1 (265 cases) results in only KSH 208 million**, showing that some days are driven by higher-value fraud rather than higher frequency.

At the weekly level, fraud remains relatively stable across Weeks 1–4, ranging from about **1,794 to 1,902 cases**, indicating continuous rather than bursty activity. However, Month 1 shows **7,931 cases and KSH 11.6 billion lost**, while Month 2 drops sharply to **282 cases and KSH 449 million**. This sharp decline is due to the **simulated structure of the PaySim dataset**, not a real-world behavioral change.

Overall, fraud is consistent over time but highly variable in impact, with certain days dominated by high-value losses and others by higher transaction counts.
